Support for DataArray.expand_dims()

fujiisoup opened the pull request:
- [x] closes #1326
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

I added a DataArray's method `expand_dims` based on the discussion in #1326 .
The proposed API is similar to `numpy.expand_dims` and slightly different from `Variables.expand_dims`, which requires whole sequences of `dims` of the result array.

My concern is that I do not yet fully understand the lazy data manipulation in xarray.
Does Variable.expand_dims do it?

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+        dim : str or sequence of str, or None.
+            Dimensions to include on the new variable.
+            dimensions are inserted with length 1.
+            If None is passed, dimension labels are automatically generated,
```
I would rather not support automatic dimension naming anywhere other than the DataArray constructor. It's error prone if `array.expand_dims()` usually makes an array with a first dimension `dim_0` and but sometimes calls it something else.

It's OK for DataArray constructors because (a) it mostly reminds users that they need to supply names and (b) usually every name is given a default value, so we don't need to worry about conflicts.

fujiisoup replied:
OK. I will remove this default argument.

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+        axis, dim = zip(*sorted(zip(axis, dim)))
+        # dims of the result array. This will be passed to Variable.expand_dims
+        all_dims = list(self.dims)
+        [all_dims.insert(a, d) for a, d in zip(axis, dim)]
```
Use a loop in cases where you throw away the result of a comprehension, e.g.,
```
for a, d in zip(axis, dim):
    all_dims.insert(a, d)
```

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+
+        # If dims includes a label of a scalar variables,
+        # it will be promoted to a 1D coordinate consisting of a single value.
+        coords = OrderedDict(self.coords)
```
Use `self._coords`, which is an OrderedDict of `xarray.Variable` objects, instead of `self.coords`

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+        # it will be promoted to a 1D coordinate consisting of a single value.
+        coords = OrderedDict(self.coords)
+        for d in dim:
+            if d in coords.keys():
```
Just check `if d in coords:`

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+        coords = OrderedDict(self.coords)
+        for d in dim:
+            if d in coords.keys():
+                coords[d] = np.atleast_1d(coords[d])
```
Instead of calling `np.atleast_1d`, use something like `Variable.expand_dims` to ensure this plays nicely with with non-numpy arrays.

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
         ds = self._to_temp_dataset().swap_dims(dims_dict)
         return self._from_temp_dataset(ds)
 
+    def expand_dims(self, dim=None, axis=None):
```
Consider implementing this on `xarray.Dataset` instead (looping over all data variables), using `_to_temp_dataset()`/`_from_temp_dataset()` to write the `DataArray` method: see `swap_dims` above for an example.

fujiisoup replied:
What is the reason to implement in `xarray.Dataset`?  Does it mean that we would also support `Dataset.expand_dims`? 

I think I need to make another design decision.

`axis` argument for `xarray.Dataset.expand_dims(dim, axis)` is somewhat problematic because `data_var`s are not necessarily the same dimension size.

It would be reasonable to raise an ValueError if any of `data_var` has smaller dimension that is required by `axis`?

shoyer replied:
> What is the reason to implement in xarray.Dataset? Does it mean that we would also support Dataset.expand_dims?

Yes, exactly.

> It would be reasonable to raise an ValueError if any of data_var has smaller dimension that is required by axis?

Yes, of course. In practice, I suspect most people will either keep the default (to expand dims at the start) or use negative integers (to expand dims at the end).

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+
+        Raises
+        -------
+        ValueError:
```
It's good form to document when functions can raise, but all of these conditions are already implied by the documentation for the `axis` argument, so I don't think you need to include this section. Essentially every Python function can raise `ValueError` when passed invalid arguments.

shoyer commented:
> The proposed API is similar to numpy.expand_dims and slightly different from Variables.expand_dims, which requires whole sequences of dims of the result array.

This looks great, but to avoid confusion, can we rename the existing `Variable.expand_dims` method? Maybe `with_dims` would be a clearer name.

> My concern is that I do not yet fully understand the lazy data manipulation in xarray.
Does Variable.expand_dims do it?

Yes, `Variable.expand_dims` is already lazy on dask arrays.

fujiisoup commented:
> Maybe with_dims would be a clearer name.
How about `reshape`? 
I think verb-name gives more intuitition about the action of the method.
The current implementation of `Variable.expand_dims` actually does this when `Ordered_dict` is passed.

fujiisoup commented:
I moved `expand_dims` to `Dataset`.

How can I add API docs? 
Only to add a single line to api.rst?

> How about `reshape`?

I changed my mind. `reshape` sounds too general. How about `append_dims` or `attach_dims` ?

shoyer commented:
> How can I add API docs?
> Only to add a single line to api.rst?

Yes, that's correct.

> I changed my mind. reshape sounds too general. How about append_dims or attach_dims ?

Indeed, I would rather not do reshape, because this method does not support arbitrary reshaping (e.g., length 100 vector-> 10x10 matrix). Instead, it assigns dimensions using array broadcasting and transpose, which unlike reshape does not ever copy data. Maybe `set_dims` would be better? I also would be OK with `attach_dims`. `append_dims` makes it sound like dimensions are always appended (they aren't).

Actually, `reshape` could indeed be a nice method to add to `Dataset`/`DataArray`, as an alternative to `stack`/`unstack`.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        # make sure any of dim does not already exist
+        for d in dim:
+            if d in self.dims:
+                raise ValueError('{dim} already exists.'.format(dim=d))
```
Include the word "dimension" so it's clear what the problem is: `'dimension {dim} already exists.'`

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+
+        # make sure there is no duplicates in axis or dims
+        if len(axis) != len(set(axis)):
+            raise ValueError('axis should not contain same values.')
```
Grammar: same values -> duplicate values

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        if len(axis) != len(set(axis)):
+            raise ValueError('axis should not contain same values.')
+        if len(dim) != len(set(dim)):
+            raise ValueError('dims should not contain same values.')
```
same -> duplicate, as above

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        variables = OrderedDict()
+        for k, v in iteritems(self._variables):
+            if k in self._coord_names:
+                if k not in dim: # Do not change coordinates
```
Switch the order of these conditions to avoid checking for a negative: it results in clearer code.

Also, [per PEP8](https://www.python.org/dev/peps/pep-0008/#inline-comments) include two spaces before any inline comments.

And per xarray terminology, call these "non-dimension coordinates": http://xarray.pydata.org/en/stable/data-structures.html#coordinates

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        expanded : same type as caller
+            This object, but with an additional dimension(s).
+        """
+        # Make sure user does not do `expand_dims(0)`
```
This comment (and most of the others below) is simply describing what the code does. It would be better to omit it.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        if isinstance(dim, int):
+            raise ValueError('dim should be str or sequence of strs or dict')
+
+        # Make axis and dims a pair of lists
```
delete

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        if axis is not None and not isinstance(axis, (list, tuple)):
+            axis = [axis]
+
+        # infer None arguments.
```
delete

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+            axis = list(range(len(dim)))
+
+        # Error checking
+        # Make sure the length of axis and dims are identical
```
delete

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        # Make sure the length of axis and dims are identical
+        if len(dim) != len(axis):
+            raise ValueError('lengths of dim and axis should be identical.')
+        # make sure any of dim does not already exist
```
delete

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+            if d in self.dims:
+                raise ValueError('{dim} already exists.'.format(dim=d))
+
+        # make sure there is no duplicates in axis or dims
```
delete

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,82 @@ def swap_dims(self, dims_dict, inplace=False):
+        if len(dim) != len(set(dim)):
+            raise ValueError('dims should not contain same values.')
+
+        # sort dim and axis, so that axis is in order.
```
Note that this is for `.insert` below.

fujiisoup commented:
Thanks for the review.

I completely forgot the negative axis. Thanks.
It is supported by my last commit.

I also changed `Variable.expand_dims` to `Variable.set_dims`.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1599,55 +1599,50 @@ def expand_dims(self, dim, axis=None):
+                    all_dims = list(v.dims)
+                    for a, d in zip(axis_pos, dim_pos):
+                        all_dims.insert(a, d)
+                    print(all_dims)
```
Remove

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1599,55 +1599,50 @@ def expand_dims(self, dim, axis=None):
+                                for a in axis]
+                    if len(axis_pos) != len(set(axis_pos)):
+                        raise ValueError('axis should not contain duplicate'
+                                         ' values.')
```
You also need to check for out of bound axes, e.g., if `result_ndim = len(v.dims) + len(axis)`, then all axes `a` must satisfy `-result_ndim <= a < result_ndim`.

This is probably easier before making axis positive so you can give an array message with the original, incorrect axis number.

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1599,55 +1599,50 @@ def expand_dims(self, dim, axis=None):
+                                         ' values.')
+                    # We need to sort them to make sure `axis` equals to the
+                    # axis positions of the result array.
+                    axis_pos, dim_pos = zip(*sorted(zip(axis_pos, dim)))
```
Your code currently zips and unzips (here) and rezips (in the loop). You can remove two of these for a small speedup.

shoyer commented on xarray/tests/test_dataarray.py:
```diff
@@ -913,6 +913,88 @@ def test_swap_dims(self):
+        with self.assertRaises(ValueError):
+            array.expand_dims(dim=['y', 'z'], axis=[2, -2])
+
+        # Working test
```
Consider breaking these into smaller test methods, e.g., `test_expand_dims_errors`. It's  easier to debug when tests each one thing.

fujiisoup commented:
Thanks for the careful review.

+ Now it raises `IndexError` if any axis is out of bounds.
   `ValueError` is OK for the other part?
+ Removed unnecessary `print` line.
+ Removed the `unzip` -> `rezip` sequences.
+ Divide `test_dataarray.test_expand_dims` into small 2 tests.

shoyer commented on xarray/core/variable.py:
```diff
@@ -706,8 +706,9 @@ def transpose(self, *dims):
 
-    def expand_dims(self, dims, shape=None):
-        """Return a new variable with expanded dimensions.
+    def set_dims(self, dims, shape=None):
```
Can you keep around `Variable.expand_dims` as an alias for `set_dims` with a deprecation warning?

e.g., see the deprecation warning for `broadcast_arrays`: https://github.com/pydata/xarray/blob/93d6963315026f87841c7cf39cc39bb78f555345/xarray/core/alignment.py#L531

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,90 @@ def swap_dims(self, dims_dict, inplace=False):
+            if d in self.dims:
+                raise ValueError(
+                            'Dimension {dim} already exists.'.format(dim=d))
+            if d in self._variables.keys() and \
```
Per PEP8, use parentheses when possible instead of explicit continuations with `\`, e.g.,
```
if (d in a_very_long_expression and
        not something_else_very_long):
```

Also, check `d in self._variables` directly instead of using `.keys()` (which is a list Python 2, making for very slow lookups).

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1575,6 +1575,90 @@ def swap_dims(self, dims_dict, inplace=False):
+                    result_ndim = len(v.dims) + len(axis)
+                    for a in axis:
+                        if a < -result_ndim or result_ndim - 1 < a:
+                            raise IndexError(
```
Either `IndexError` or `ValueError` would be appropriate here. NumPy has a new error type `np.AxisError` but it's only in the latest, unreleased version of NumPy.

fujiisoup commented:
I added a section to `reshaping.rst`, which shows an example (default) usage of `expand_dims` and `squeeze`.

shoyer commented:
This looks great. Thank you @fujisoup!

fujiisoup commented:
Oops. Found a typo in reshaping.rst. I'll send another PR.
